contextily/tile.py

```diff
@@ -205,6 +209,125 @@ def _fetch_tile(tile_url, wait, max_retries):
+                  Image as a 3D array (h, w, b) of RGB values (e.g. as
+                  returned from `contextily.bounds2img`)
+    ext         : tuple
+                  Bounding box [minX, maxX, minY, maxY] of the returned (warped)
+                  image
+    '''
+    h, w, b = img.shape
+    # --- https://mapbox.github.io/rasterio/quickstart.html#opening-a-dataset-in-writing-mode
+    minX, maxX, minY, maxY = ext
+    x = np.linspace(minX, maxX, w)
+    y = np.linspace(minY, maxY, h)
+    resX = (x[-1] - x[0]) / w
+    resY = (y[-1] - y[0]) / h
+    transform = from_origin(x[0] - resX / 2,
+                            y[-1] + resY / 2, resX, resY)
+    # --- 
+    w_img, vrt = _warper(img.transpose(2, 0, 1), 
+                         transform, 
+                         'EPSG:3857', t_crs,
+                         resampling)
+    # --- 
+    extent = vrt.bounds.left, vrt.bounds.right, \
+             vrt.bounds.bottom, vrt.bounds.top
+    return w_img.transpose(1, 2, 0), extent
+
+
+def warp_img_transform(img, transform, 
+                       s_crs, t_crs,
+                       resampling=Resampling.bilinear):
+    '''
+    Reproject (warp) an `img` with a given `transform` and `s_crs` into a
+    different `t_crs`
+
+    NOTE: this method works well with rasterio's `.read()` approach to
+          raster's dimensions (b, h, w)
+    ...
+
+    Arguments
+    ---------
+    img         : ndarray
+                  Image as a 3D array (b, h, w) of RGB values (e.g. as
+                  returned from rasterio's `.read()` method)
+    transform   : affine.Affine
+                  Transform of the input image as expressed by `rasterio` and
+                  the `affine` package
+    s_crs       : str/CRS
+                  Source CRS in which `img` is passed, expressed in any format
+                  permitted by rasterio.
+    t_crs       : str/CRS
+                  Target CRS, expressed in any format permitted by rasterio.
+    resampling  : <enum 'Resampling'>
+                  [Optional. Default=Resampling.bilinear] Resampling method for
+                  executing warping, expressed as a `rasterio.enums.Resampling
+                  method
+
+    Returns
+    -------
+    w_img       : ndarray
+                  Warped image as a 3D array (b, h, w) of RGB values (e.g. as
+                  returned from rasterio's `.read()` method)
+    w_transform : affine.Affine
+                  Transform of the input image as expressed by `rasterio` and
+                  the `affine` package
+    '''
+    w_img, vrt = _warper(img, transform, 
+                         s_crs, t_crs,
+                         resampling)
+    return w_img, vrt.transform
+
+
+def _warper(img, transform, 
+            s_crs, t_crs,
+            resampling):
+    '''
+    Warp an image returning it as a virtual file
+    '''
+    b, h, w = img.shape
+    with MemoryFile() as memfile:
+        with memfile.open(driver='GTiff', height=h, width=w, \
```

Comment:
Is there a reason you use this MemoryFile approach instead of using the `rasterio.warp.reproject` function? (https://rasterio.readthedocs.io/en/latest/api/rasterio.warp.html) 
Although there you need to construct an empty result array first manually to pass to that function, so that might make it a bit more contrived to use.

Reply:
I actually hadn't seen that approach, it might be newer than the implementation here? It'd need some detailed looking. In any case, we could drop-in replace later if it's deemed preferable? The current approach also doesn't require writing anything to disk and, in tests, it seems to perform OK. Leave for now and we can test alternatives later?

Reply:
Yes, sure, fine to leave it for later (I was just wondering, it's this approach that I used a while ago if I remember correctly)

Reply:
Should we turn it into an issue to not forget about it? If someone has time, it'd be good to explore the benefits that might bring (maybe more performant/efficient?). Set as #88